pdf/pdfselection.cpp

```diff
@@ -0,0 +1,439 @@
+}
+void PDFSelection::setStart(const QPointF &pt)
+{
+    int iPage, iBox;
+    d->textBoxId(pt, PDFSelection::Private::After, &iPage, &iBox);
+    if (iPage < 0 || iBox < 0 || (iPage == d->iPageStart && iBox == d->iBoxStart))
+        return;
+    if (iPage > d->iPageStop || (iPage == d->iPageStop && iBox > d->iBoxStop))
+        return;
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(iPage);
+    QRectF box = boxes[iBox].first;
+
+    int nBoxes = d->sliceCount(iPage, iBox, d->iPageStart, d->iBoxStart);
+    if (nBoxes > 0) {
+        beginInsertRows(QModelIndex(), 0, nBoxes - 1);
+        d->iPageStart = iPage;
+        d->iBoxStart = iBox;
+        endInsertRows();
+    } else {
+        beginRemoveRows(QModelIndex(), 0, -nBoxes - 1);
+        d->iPageStart = iPage;
+        d->iBoxStart = iBox;
+        endRemoveRows();
+    }
+    emit countChanged();
+    emit textChanged();
+
+    d->start = QPointF(box.x(), box.y() + box.height() / 2.);
+    emit startChanged();
+}
+
+QPointF PDFSelection::stop() const
+{
+    if (!d->canvas || d->iPageStop < 0)
+        return QPointF();
+    
+    return d->canvas->fromPageToItem(d->iPageStop, d->stop);
+}
+void PDFSelection::setStop(const QPointF &pt)
+{
+    int iPage, iBox;
+    d->textBoxId(pt, PDFSelection::Private::Before, &iPage, &iBox);
+    if (iPage < 0 || iBox < 0 || (iPage == d->iPageStop && iBox == d->iBoxStop))
+        return;
+    if (iPage < d->iPageStart || (iPage == d->iPageStart && iBox < d->iBoxStart))
+        return;
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(iPage);
+    QRectF box = boxes[iBox].first;
+
+    int count_ = count(); 
+    int nBoxes = d->sliceCount(d->iPageStop, d->iBoxStop, iPage, iBox);
+    if (nBoxes > 0) {
+        beginInsertRows(QModelIndex(), count_, count_ + nBoxes - 1);
+        d->iPageStop = iPage;
+        d->iBoxStop = iBox;
+        endInsertRows();
+    } else {
+        beginRemoveRows(QModelIndex(), count_ + nBoxes, count_ - 1);
+        d->iPageStop = iPage;
+        d->iBoxStop = iBox;
+        endRemoveRows();
+    }
+    emit countChanged();
+    emit textChanged();
+
+    d->stop = QPointF(box.x() + box.width(), box.y() + box.height() / 2.);
+    emit stopChanged();
+}
+
+void PDFSelection::selectAt(const QPointF &pt)
+{
+    int iPage, iBox;
+    d->textBoxId(pt, PDFSelection::Private::At, &iPage, &iBox);
+    if (iPage < 0 || iBox < 0) {
+        beginResetModel();
+        d->iPageStart = d->iPageStop = -1;
+        d->iBoxStart = d->iBoxStop = -1;
+        endResetModel();
+        emit countChanged();
+        emit textChanged();
+        return;
+    }
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(iPage);
+    QRectF box = boxes[iBox].first;
+    
+    beginResetModel();
+    d->iPageStart = -1;
+    d->iBoxStart = -1;
+    d->iPageStop = -1;
+    d->iBoxStop = -1;
+    endResetModel();
```

Comment:
Well, here, I want to set a new coontent for the model, so do a clear and a new insert.

If I don't reset the model before inserting, doing a selection works fine, but selecting another word after make the previous selection still existing.